avr-hal-generic/src/spi.rs

```diff
@@ -64,220 +66,258 @@ impl Default for Settings {
-            }
+    /// Reconfigure the SPI peripheral after initializing
+    pub fn reconfigure(&mut self, settings: Settings) -> nb::Result<(), crate::void::Void> {
+        // wait for any in-flight writes to complete
+        self.p.raw_flush()?;
+        self.p.raw_setup(&settings);
+        Ok(())
+    }
 
-            /// Check if write flag is set, and return a WouldBlock error if it is not.
-            fn flush(&mut self) -> $crate::nb::Result<(), $crate::void::Void> {
-                if self.is_write_in_progress {
-                    if self.peripheral.spsr.read().spif().bit_is_set() {
-                        self.is_write_in_progress = false;
-                    } else {
-                        return Err($crate::nb::Error::WouldBlock);
-                    }
-                }
-                Ok(())
-            }
+    /// Disable the SPI device and release ownership of the peripheral
+    /// and pins.  Instance can no-longer be used after this is
+    /// invoked.
+    pub fn release(
+        mut self,
+        cs: ChipSelectPin<CSPIN>,
+    ) -> (
+        SPI,
+        port::Pin<port::mode::Output, SCLKPIN>,
+        port::Pin<port::mode::Output, MOSIPIN>,
+        port::Pin<port::mode::Input, MISOPIN>,
+        CSPIN,
+    ) {
+        self.p.raw_release();
+        (self.p, self.sclk, self.mosi, self.miso, cs.0)
+    }
+}
+
+/// FullDuplex trait implementation, allowing this struct to be provided to
+/// drivers that require it for operation.  Only 8-bit word size is supported
+/// for now.
+impl<H, SPI, SCLK, MOSI, MISO, CS> spi::FullDuplex<u8> for Spi<H, SPI, SCLK, MOSI, MISO, CS>
+where
+    SPI: SpiOps<H, SCLK, MOSI, MISO, CS>,
+{
+    type Error = void::Void;
 
+    /// Sets up the device for transmission and sends the data
+    fn send(&mut self, byte: u8) -> nb::Result<(), Self::Error> {
+        self.p.raw_flush()?;
+        self.p.raw_write(byte);
+        Ok(())
+    }
+
+    /// Reads and returns the response in the data register
+    fn read(&mut self) -> nb::Result<u8, Self::Error> {
+        self.p.raw_flush()?;
+        Ok(self.p.raw_read())
+    }
+}
+
+/// Default Transfer trait implementation. Only 8-bit word size is supported for now.
+impl<H, SPI, SCLK, MOSI, MISO, CS> hal::blocking::spi::transfer::Default<u8>
+    for Spi<H, SPI, SCLK, MOSI, MISO, CS>
+where
+    SPI: SpiOps<H, SCLK, MOSI, MISO, CS>,
+{
+}
+
+/// Default Write trait implementation. Only 8-bit word size is supported for now.
+impl<H, SPI, SCLK, MOSI, MISO, CS> hal::blocking::spi::write::Default<u8>
+    for Spi<H, SPI, SCLK, MOSI, MISO, CS>
+where
+    SPI: SpiOps<H, SCLK, MOSI, MISO, CS>,
+{
+}
+
+/// Implement traits for a SPI interface
+#[macro_export]
+macro_rules! impl_spi {
+    (
+        hal: $HAL:ty,
+        peripheral: $SPI:ty,
+        sclk: $sclkpin:ty,
+        mosi: $mosipin:ty,
+        miso: $misopin:ty,
+        cs: $cspin:ty,
+    ) => {
+        static mut IS_WRITE_IN_PROGRESS: bool = false;
```

Comment:
This has multiple issues:

1. It means you cannot use the `impl_spi!{}` macro multiple times in the same module due to how macros work.
2. It is an unsynchronized global variable so it can very easily lead to memory races.  That's no good.

I think it's the wrong approach to have this here.  This bitflag should be stored in the `Spi<>` higher level type and the lower level APIs (`SpiOps`) should be designed such that the flag is checked and set in the `Spi<>` methods.  As an example, the following could work:

```rust
impl SpiOps for ... {
    fn raw_write(&mut self, byte: u8) {
        self.spdr.write(|w| unsafe { w.bits(byte) });
    }

    fn raw_check_iflag(&mut self) -> bool {
        self.spsr.read().spif().bit_is_set()
    }
}

impl Spi<...> {
    fn write_internal(&mut self, byte: u8) {
        self.is_write_in_progress = true;
        self.p.raw_write(byte);
    }

    fn flush_internal(&mut self) -> nb::Result<(), void::Void> {
        if self.is_write_in_progress {
            if self.p.raw_check_iflag() {
                self.is_write_in_progress = false;
            } else {
                return Err(nb::Error::WouldBlock);
            }
        }
        Ok(())
    }
}

impl FullDuplex for Spi<...> {
    fn send(&mut self, byte: u8) -> nb::Result<(), Self::Error> {
        self.flush_internal()?;
        self.write_internal(byte);
        Ok(())
    }

    fn read(&mut self) -> nb::Result<u8, Self::Error> {
        self.flush_internal()?;
        Ok(self.p.raw_read())
    }
}
```

Reply:
Yea, thanks, I didn't love this solution but was just trying to get something to compile.  Your solution is better, I think.